system/gsettings.py

```diff
@@ -0,0 +1,258 @@
+    return proc_environ(pid, ENV_DBUS)
+
+def new_session():
+    """Create a new dbus session"""
+    dbus_process = subprocess.Popen('dbus-launch', stdout=subprocess.PIPE)
+    for env_string in dbus_process.stdout.readlines():
+        env_var, env_value = env_string.split('=', 1)
+        os.environ[env_var] = env_value
+    return os.environ['DBUS_SESSION_BUS_PID'].replace('\n', '')
+
+
+class Gsettings(object):
+    def __init__(self):
+        # Determine the session pid.
+        self.session_pid = -1
+        self._cleanup = False
+
+        for procfile in os.listdir('/proc'):
+            if procfile.isdigit():
+                if proc_name(procfile) in SESSION_MANAGERS:
+                    self.session_pid = procfile
+
+        osuser = os.getuid()
+        if self.session_pid != -1:
+            sessionuser = proc_owner(self.session_pid)
+        else:
+            sessionuser = osuser
+
+        # Setup the environment and connect to the bus.
+        if self.session_pid != -1 and osuser == sessionuser:
+            os.environ[ENV_DBUS] = bus_address(self.session_pid)
+        else:
+            self.session_pid = new_session()
+            self._cleanup = True
+
+        self.bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
+
+    def get_value(self, schema, key):
+        """ Read the value of the specified key """
+        # ':' implies relocatable schemas
+        if ':' in schema:
+            schema, path = schema.split(':', 1)
+            schemas = Gio.Settings.list_relocatable_schemas()
+        else:
+            path = None
+            schemas = Gio.Settings.list_schemas()
+
+        # Gio.Settings does not fail gracefully, making
+        # use of abort when an exception occurs.  Thus,
+        # careful checking of values is done before
+        # commiting to get_value.
+        if schema not in schemas:
+            raise ValueError('{0} is not a valid schema'.format(schema))
+
+        if path is None:
+            settings = Gio.Settings(schema=schema, path=path)
+        else:
+            settings = Gio.Settings.new_with_path(schema_id=schema, path=path)
+
+        if key not in Gio.Settings.list_keys(settings):
+            raise ValueError('{0} is not valid in the schema: {1}'.format(key, schema))
+        return settings.get_value(key)
+
+    def set_value(self, schema, key, value):
+        """ Write a value to the specified key """
+        if ':' in schema:
+            schema, path = schema.split(':', 1)
+        else:
+            path = None
+
+        # No need to check values passed as they were
+        # already checked in the get_value function.
+        settings = Gio.Settings(schema=schema, path=path)
+        settings.set_value(key, value)
+
+    def cleanup(self):
+        """ Class cleanup """
+        if self._cleanup:
+            # Stop the dbus daemon if one was started.
+            time.sleep(1)
+            os.kill(int(self.session_pid), signal.SIGTERM)
```

Comment:
shouldn't we verify that the PID is still the one of the dbus daemon ? if it did crash, this would send a SIGTERM to a random process somewhere, which is likely a unattended side effect.